code/modules/mob/living/simple_animal/friendly/mule_bot/mule_bot.dm

```diff
@@ -0,0 +1,202 @@
+	flags_atom = BUMP_ATTACKABLE //bump attack may not work due to pass flags
+	light_color =  "#3262db"
+	light_range = 2
+	light_power = 1
+	light_on = FALSE
+	//TODO to make this  hud actuely work
+	hud_possible = list(MACHINE_HEALTH_HUD)
+	maxHealth  = 200
+	var/mutable_appearance/face_overlay
+	var/obj/item/mule_module/installed_module
+	var/obj/item/clothing/head/hat
+	var/mutable_appearance/hat_overlay
+
+
+/mob/living/simple_animal/mule_bot/Initialize()
+	. = ..()
+	face_overlay = emissive_appearance(icon, "kerfus_face")
+	update_icon()
+
+/mob/living/simple_animal/mule_bot/proc/try_link(obj/item/remote/R)
+	if(R)
+		AddComponent(/datum/component/ai_controller, /datum/ai_behavior/mule_bot, R)
+
+/mob/living/simple_animal/mule_bot/update_overlays()
+	. = ..()
+	. += list(hat_overlay,face_overlay,installed_module?.mod_overlay)
+
+
+/mob/living/simple_animal/mule_bot/attackby(obj/item/I, mob/living/user, def_zone)
+	//give it a funny hat, could be turned into an actuel proc though
+	if(istype(I,/obj/item/clothing/head))
+		if(hat)
+			hat.forceMove(src.loc)
+		var/obj/item/clothing/head/new_hat = I
+		I.forceMove(src)
+		user?.temporarilyRemoveItemFromInventory(I)
+		hat = new_hat
+		hat_overlay = mutable_appearance(new_hat.get_worn_icon_file("Human",slot_head_str), new_hat.get_worn_icon_state(slot_head_str), HEAD_LAYER, FLOAT_PLANE)
+		hat_overlay.pixel_y -= -5
+		update_icon()
+		return
+	if(istype(I,/obj/item/mule_module))
+		apply_module(I,user)
+		return
+	. = ..()
+
+/mob/living/simple_animal/mule_bot/welder_act(mob/living/user, obj/item/I)
+	var/repair_time = 1 SECONDS
+	if(src == user)
+		repair_time *= 3
+
+	user.visible_message(span_notice("[user] starts to fix some of the dents on [src]."),\
+		span_notice("You start fixing some of the dents on [src]"))
+
+	add_overlay(GLOB.welding_sparks)
+	while(do_after(user, repair_time, TRUE, src, BUSY_ICON_BUILD) && I.use_tool(volume = 50, amount = 2))
+		user.visible_message(span_warning("\The [user] patches some dents on [src]."), \
+			span_warning("You patch some dents on \the [src]."))
+		if(src.heal_limb_damage(15,15, robo_repair = FALSE, updating_health = TRUE))
+			UpdateDamageIcon()
+		if(!I.tool_use_check(user, 2))
+			break
+		if(!src.bruteloss)
+			balloon_alert(user, "Dents fully repaired.")
+			break
+	cut_overlay(GLOB.welding_sparks)
+	return TRUE
+
+/mob/living/simple_animal/mule_bot/crowbar_act(mob/living/user, obj/item/I)
+	remove_module()
+	return TRUE
+
+
+/mob/living/simple_animal/mule_bot/proc/swap_module(obj/item/mule_module/mod, mob/user)
+	remove_module(user, FALSE)
+	apply_module(mod,user, FALSE)
+	update_icon()
+
+/mob/living/simple_animal/mule_bot/proc/apply_module(obj/item/mule_module/mod, mob/user, update_icon = TRUE)
+	if(mod.apply(src))
+		to_chat(user,span_notice("You succesfully installed [mod]"))
+		user?.temporarilyRemoveItemFromInventory(mod)
+	else
+		to_chat(user,span_warning("You failed to installed [mod]"))
+	if(update_icon)
+		update_icon()
```

Comment:
idk why you keep wrapping code blocks used in one place into their own procs, it's pointless and just makes the code messier

Reply:
I like the idea of calling a high level proc that makes sure the lower level stuff is done. While also having ability to call proc to skip certain stuff (say you make the bot spawn with module X you can just call apply_module instead of having to remove/check to remove the old one